panic_test.go

```diff
@@ -111,3 +111,68 @@ func funcWithBadPanic() {
 func funcWithBadErr() error {
 	return errors.New("awful")
 }
+
+func TestPanicChain(t *testing.T) {
+	ch0 := make(chan error, 1)
+	ch1 := make(chan error, 1)
+
+	doMischief(ch1)
+	doMoreMischief(ch0, ch1)
+
+	select {
+	case err := <-ch0:
+		require.Error(t, err)
+		output := fmt.Sprintf("%+v", err)
+		require.Contains(t, output, "mischiefProper", output)
+		require.Contains(t, output, "mischiefAsPanic", output)
+		require.Contains(t, output, "doMischief", output)
+		require.Contains(t, output, "handleMischief", output)
+		require.Contains(t, output, "doMoreMischief", output)
+		t.Log(output)
+	case <-time.After(time.Second):
+		require.Fail(t, "expected error")
+	}
+}
+
+func doMoreMischief(ch0 chan error, ch1 chan error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err, ok := ErrorFromPanic(e)
+			if ok {
+				ch0 <- Decorate(err, "hop 2")
+				return
+			}
+		}
+		ch0 <- AssertionFailed.New("test failed")
+	}()
+
+	handleMischief(ch1)
+}
+
+func handleMischief(ch chan error) {
+	err := <-ch
+	Panic(Decorate(err, "handle"))
+}
+
+func doMischief(ch chan error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err, ok := ErrorFromPanic(e)
+			if ok {
+				ch <- Decorate(err, "hop 1")
```

Comment:
Why are you using a channel to return value from this function? Wouldn't it be easier to just return something? Yes, I see `defer` here. You will need named return apparently.

Reply:
Fair point, it is a test half complete. The idea is to make "hops" as different in stack trace as possible - and so I want to spawn some goroutines. 
It all comes together via a simple write-read chain, and test itself has a timeout, so I don't see much of a danger here.